stackslib/src/chainstate/nakamoto/mod.rs

```diff
@@ -239,6 +247,462 @@ lazy_static! {
+        &mut self,
+        tip: &StacksBlockId,
+        tenure_id_consensus_hash: &ConsensusHash,
+    ) -> Result<Option<u64>, DBError> {
+        Ok(self
+            .get_indexed(
+                tip,
+                &nakamoto_keys::coinbase_height(tenure_id_consensus_hash),
+            )?
+            .map(|height_str| nakamoto_keys::parse_u64(&height_str))
+            .flatten())
+    }
+
+    fn get_ongoing_tenure_id(
+        &mut self,
+        tip: &StacksBlockId,
+    ) -> Result<Option<NakamotoTenureId>, DBError> {
+        Ok(self
+            .get_indexed(tip, nakamoto_keys::ongoing_tenure_id())?
+            .map(|id_str| nakamoto_keys::parse_tenure_id_value(&id_str))
+            .flatten())
+    }
+
+    fn get_highest_block_id_in_tenure(
+        &mut self,
+        tip: &StacksBlockId,
+        tenure_id_consensus_hash: &ConsensusHash,
+    ) -> Result<Option<StacksBlockId>, DBError> {
+        Ok(self
+            .get_indexed(
+                tip,
+                &nakamoto_keys::highest_block_in_tenure(tenure_id_consensus_hash),
+            )?
+            .map(|id_str| nakamoto_keys::parse_block_id(&id_str))
+            .flatten())
+    }
+
+    fn get_block_found_tenure_id(
+        &mut self,
+        tip: &StacksBlockId,
+        tenure_id_consensus_hash: &ConsensusHash,
+    ) -> Result<Option<NakamotoTenureId>, DBError> {
+        Ok(self
+            .get_indexed(
+                tip,
+                &nakamoto_keys::block_found_tenure_id(tenure_id_consensus_hash),
+            )?
+            .map(|id_str| nakamoto_keys::parse_tenure_id_value(&id_str))
+            .flatten())
+    }
+
+    fn is_tenure_finished(
+        &mut self,
+        tip: &StacksBlockId,
+        tenure_id_consensus_hash: &ConsensusHash,
+    ) -> Result<Option<bool>, DBError> {
+        if self
+            .get_tenure_start_block_id(tip, tenure_id_consensus_hash)?
+            .is_none()
+        {
+            // tenure not started
+            return Ok(None);
+        }
+        if self
+            .get_indexed(
+                tip,
+                &nakamoto_keys::finished_tenure_consensus_hash(tenure_id_consensus_hash),
+            )?
+            .is_none()
+        {
+            // tenure has started, but is not done yet
+            return Ok(Some(false));
+        }
+
+        // tenure started and finished
+        Ok(Some(true))
+    }
+
+    fn get_parent_tenure_consensus_hash(
+        &mut self,
+        tip: &StacksBlockId,
+        tenure_id_consensus_hash: &ConsensusHash,
+    ) -> Result<Option<ConsensusHash>, DBError> {
+        Ok(self
+            .get_indexed(
+                tip,
+                &nakamoto_keys::parent_tenure_consensus_hash(tenure_id_consensus_hash),
+            )?
+            .map(|ch_str| nakamoto_keys::parse_consensus_hash(&ch_str))
+            .flatten())
+    }
+
+    fn sqlite(&self) -> &Connection {
+        self.conn()
+    }
+}
+
+impl StacksHandle for StacksDBTx<'_> {
```

Comment:
The implementation of `StacksHandle` for StacksDBTx and StacksDBConn look identical. If the trait instead required `get_indexed()`, all of this duplication could be removed by providing default implementations for these functions in the `StacksHandle` trait.

Reply:
Yup, that's a better approach